feat: type guard chaining

flisboac opened the pull request:
This PR adds a new functionality: guard chaining.

With a guard chain wrapper, it's possible to create conjunctions or disjunctions of type guards that statically map to intersections or unions of types, respectively.

For example, given `isNull` and `isUndefined`, one can create a new `TypeGuard<null | undefined>` with the idiom `guard(isNull).and(isUndefined)`.

My motivation is to allow for type guard composition, and in the future, use this functionality together with others that are perhaps closely related (and that I still plan to develop and make PRs for, e.g. non-classy array filtering, type enforcers).

Please see the README and __tests__ for more information.

SimonAlling commented:
Wow, a PR! I'll have a look as soon as possible!

SimonAlling commented on README.md:
```diff
@@ -136,6 +136,29 @@ xs.filter(isFoo).forEach(x => x.baz());
+
+`guard` returns a wrapper type guard just like `is` or `isLike`, but exposes two additional properties to the type guard itself:
+
+- `guard(wrappedGuard).or(anotherGuard)` returns a new type guard that checks a value for either `wrappedGuard` or `anotherGuard`. In TypeScript parlance, `or` returns a type union-based type guard; and
```
```suggestion
- `guard(wrappedGuard).or(anotherGuard)` is a type guard that checks a value for either `wrappedGuard` or `anotherGuard`. In TypeScript parlance, `or` returns a type union-based type guard; and
```

SimonAlling commented on README.md:
```diff
@@ -136,6 +136,29 @@ xs.filter(isFoo).forEach(x => x.baz());
+`guard` returns a wrapper type guard just like `is` or `isLike`, but exposes two additional properties to the type guard itself:
+
+- `guard(wrappedGuard).or(anotherGuard)` returns a new type guard that checks a value for either `wrappedGuard` or `anotherGuard`. In TypeScript parlance, `or` returns a type union-based type guard; and
+- `guard(wrappedGuard).and(anotherGuard)` returns a new type guard that checks value for both `wrappedGuard` and `anotherGuard`. In TypeScript parlance, `and` returns a type intersection-based type guard. In general, `and` is not available for primitive type-based type guards.
```
```suggestion
- `guard(wrappedGuard).and(anotherGuard)` is a type guard that checks a value for both `wrappedGuard` and `anotherGuard`. In TypeScript parlance, `and` returns a type intersection-based type guard. In general, `and` is not available for primitive type-based type guards.
```

SimonAlling commented on src/is.ts:
```diff
@@ -107,26 +161,67 @@ export function is<T>(type: Classy<T>): TypeGuard<T> {
+ * either by union or intersection.
+ *  
+ * `guard.or` creates a union of both type checks, e.g.
+ * `guard(is(Animal)).or(isUndefined) => TypeGuard<Animal | undefined>`.
```
```suggestion
 * `guard(is(Animal)).or(isUndefined) : TypeGuard<Animal | undefined>`.
```

SimonAlling commented on src/is.ts:
```diff
@@ -107,26 +161,67 @@ export function is<T>(type: Classy<T>): TypeGuard<T> {
+ * that with a `guard` wrapper it's possible to chain type guards together,
+ * either by union or intersection.
+ *  
+ * `guard.or` creates a union of both type checks, e.g.
```
```suggestion
 * `.or` creates a union type guard, e.g.
```

SimonAlling commented on src/is.ts:
```diff
@@ -107,26 +161,67 @@ export function is<T>(type: Classy<T>): TypeGuard<T> {
+ * `guard.or` creates a union of both type checks, e.g.
+ * `guard(is(Animal)).or(isUndefined) => TypeGuard<Animal | undefined>`.
+ * 
+ * `guard.and` creates an intersection of both type checks,
```
```suggestion
 * `.and` creates an intersection type guard, e.g.
```

SimonAlling commented on src/is.ts:
```diff
@@ -107,26 +161,67 @@ export function is<T>(type: Classy<T>): TypeGuard<T> {
+ * `guard(is(Animal)).or(isUndefined) => TypeGuard<Animal | undefined>`.
+ * 
+ * `guard.and` creates an intersection of both type checks,
+ * e.g. `guard(is(Animal)).and(is(Human)) => TypeGuard<Animal & Human>`.
```
```suggestion
 * `guard(is(Animal)).and(is(Human)) : TypeGuard<Animal & Human>`.
```

SimonAlling commented on src/is.ts:
```diff
@@ -107,26 +161,67 @@ export function is<T>(type: Classy<T>): TypeGuard<T> {
+ * @param guard An object to use as reference for the type guard.
+ *
+ * @return A wrapped chainable type guard, that will perform exactly the
+ * same check as its.
```
```suggestion
 * same check as `guard`.
```

SimonAlling commented:
Hi again!

This sudden influx of contributions has caught me by surprise. It's absolutely amazing that you have put all this effort into improving `ts-type-guards`. :D

In general, I like the idea of chainable type guards – especially the fact that a builder is also a type guard itself. Also, I must take a moment to express my appreciation towards your clear, readable and careful style of programming. :+1: 

I have some questions and some change requests; I'll start with the latter:

  * The ordered list in the readme is intentionally numbered using `1.`, because otherwise, adding an item results in a much less informative diff. And it's not related to this commit anyway.
  * I'm not very fond of the name `guard` for the function that creates a builder. I suggest `chainable` or `wrap` instead.
  * Assuming said name change, I would like the library to export these:
    ```typescript
    export const both = chainable;
    export const either = chainable;
    ```
    That way, we get `both(is(A)).and(is(B))` and `either(isNull).or(isUndefined)`, which I find more readable. (Then perhaps it shouldn't export `chainable` at all. What do you think?)
  * I think `isNullLike` belongs to its own commit. (In fact, I have implemented it locally as `isNothing` – and an `isSomething` to go with it.)

I also have some change requests regarding prose, which I have added as suggestions in GitHub's GUI.

Then to the questions:

  * `.or` is obviously useful, but what is the purpose of `.and`? I haven't been able to conceive a scenario where it would be useful. You mentioned (and it makes sense) that it can't be used for primitive type guards, and interfaces can't be used with type guards at all. The only example I could imagine is something like this:
    ```typescript
    const isAnimalAndLion = both(is(Animal)).and(is(Lion));
    if (isAnimalAndLion(simba)) {
        simba.dna; // from Animal
        simba.mane; // from Lion
    }
    ```
    But when would I want such a type guard? Can't I just use `is(Lion)`?
  * When is it an advantage that a chainable guard is itself a type guard?
  * I feel confused by the new implementation of `is`:
    ```typescript
    const isFive = is(5);
    ```
    This wasn't allowed before, because I thought it wasn't reasonable to use the value `5` as a type. Furthermore, `is` is intended to mimic `instanceof`, which throws an error if the right hand side is a primitive value. (Of course we'd rather catch that error at compile time than throw it at runtime.)
  * I would also like to ask if you could explain the design choices behind all the new types and how they fit together. I'm thinking of `PrimitiveTypeGuard`, `NonPrimitiveTypeGuard`, `UnionTypeGuard`, `ChainableTypeGuard` and `FullyChainableTypeGuard`.

SimonAlling commented:
I just pushed some commits which I had forgotten to push. One of them adds `isNothing` and `isSomething`; the rest is just fixing vulnerabilities and upgrading the TypeScript version. I think you should be able to merge those changes fairly easily. Otherwise I'll be happy to help.

By the way, I think your PR should not edit `package-lock.json`. It seems like the changes you made to it were actually made by `npm install`, so I feel obliged to mention `git add -p` – I always use that flag myself! :)

And one last thing: I've been using commit messages starting with a capital letter up until now, so I'd suggest we stick to that convention.

flisboac commented:
Hello, @SimonAlling !

> This sudden influx of contributions has caught me by surprise. It's absolutely amazing that you have put all this effort into improving `ts-type-guards`. :D

I'm sorry, really! This is a lot of PRs, I admit! :D

In fact, I had a lot of ideas that should work together, but I thought that introducing them all at once would make for a harder reviewing process. I thought it would be better to isolate individual ideas and make a PR for each of them, and take a bottom-up approach (kinda). Depending on which and how many of them are accepted, I can introduce one or two more PRs to make them work better together, or change my course of action.

> I also have some change requests regarding prose, which I have added as suggestions in GitHub's GUI.

I took a look, and I'll accept all of them.

> * The ordered list in the readme is intentionally numbered using `1.`, because otherwise, adding an item results in a much less informative diff. And it's not related to this commit anyway.

Age-old mistake of adding files without checking if the editor made anything funny with it without your consent... I didn't even edit those lines myself. I'm sorry for not paying more attention!

> * I'm not very fond of the name `guard` for the function that creates a builder. I suggest `chainable` or `wrap` instead.
> * Assuming said name change, I would like the library to export these:
>   ```ts
>   export const both = chainable;
>   export const either = chainable;
>   ```
>   
>   That way, we get `both(is(A)).and(is(B))` and `either(isNull).or(isUndefined)`, which I find more readable. (Then perhaps it shouldn't export `chainable` at all. What do you think?)

`wrap` is certainly a better name. I accept that bikeshed, wholeheartedly. `chainable` is good too, but perhaps it's a tad bit too many letters. It's your call, I don't mind a bit! :D

`both` and `either` are also good ideas. I intended for `guard` to be a generic entrypoint for any type guard, regardless of the guarded type (i.e. primitive or not). `guard` is also typed so that it doesn't allow nonsensical chaining, e.g. `guard(isString).and(is(Animal))`. This way, the user would not need to make a decision about whether `both` or `either` is applicable, and could use it in e.g. an algorithm as a generic entrypoint for guard chaining -- although I admit that, in some cases, using `both`  or `either` explicitly may be a cleaner option.

The exports of  `both`  and `either` are exactly the kind of thing I'd suggest, if you had not suggested them yourself.

In time: my initial idea would be to extend the guard function namespace enough so that the library's default guards could be naturally chained (and conform to the type chaining concept purely by interface matching), e.g.:

```typescript
function isNothing(value: any): value is undefined | null {
   return value === undefined || value === null;
}
namespace isNothing {
   export const or = <T>(guard: TypeGuard<T>) => guard(isNothing).or(guard);
}
```

Then you could simply do:

```typescript
if (isNothing.or(isString)(x)) {
   // Now we know x is either not provided or is a string
}
```

But I thought the change to be a little bit too intrusive for a first PR. But if you agree with that, I still hold the opinion that it can be a good addition -- and in fact, I find it more intuitive, easier to explain and more pleasing to the eye than using a "middleware". I can make a new PR for this if the type chaining PR is merged.

> * I think `isNullLike` belongs to its own commit. (In fact, I have implemented it locally as `isNothing` – and an `isSomething` to go with it.)

I agree. Will remove it.

> * `.or` is obviously useful, but what is the purpose of `.and`? I haven't been able to conceive a scenario where it would be useful. You mentioned (and it makes sense) that it can't be used for primitive type guards, and interfaces can't be used with type guards at all. The only example I could imagine is something like this:
>   ```ts
>   const isAnimalAndLion = both(is(Animal)).and(is(Lion));
>   if (isAnimalAndLion(simba)) {
>       simba.dna; // from Animal
>       simba.mane; // from Lion
>   }
>   ```
>   
>   
>   But when would I want such a type guard? Can't I just use `is(Lion)`?

`both(is(Animal)).and(is(Lion))` looks silly, but this is naturally expected, because you're thinking in terms of classes and strict prototypical inheritance.

When I introduced the chaining concept, my intention was to validate a fetched value against a series of possible use-cases -- mostly, non-primitive (i.e. object) types. `ts-type-guards` is very much oriented towards primitives and classy types, but I had the intention of introducing interface checking in the future (I'm still preparing the PR for that), for when you don't have a base class for a value, or can't/don't want to instantiate a class for a trivial data type but still want to make checks against its interface (e.g. abstract classes without concrete classes).

Although I admit there may not be much uses for `and`, by making it available I can guarantee completeness in terms of typing. TypeScript provides unions and intersections, so by having `and` we can avoid surprises for users of the library. I also  tried to model and automate that typing as guard chains, so that the user would not need to make a custom variable/function everytime a guard union or intersection is needed.

This will be a difficult defense to make, but just to illustrate, here it goes! Suppose you use two different libraries as part of your React project. My imaginary custom component's `props` must have both [`WithStyles`](https://material-ui.com/guides/typescript/#augmenting-your-props-using-withstyles) and [`RouteComponentProps`](https://ts-react-boilerplate.js.org/docs/COMPLETING.html), as well as its own properties. My custom Props type would be something like:

```typescript
import { styles, PublicMatchParams } from './my-local-modules'
// Other imports

interface Properties { ... }

// All of those are interfaces
type Props = 
    & Properties
    & WithStyles<typeof styles>
    & RouteComponentProps<PublicMatchParams>
    ;
```

Considering that each composing type has its own guard somewhere in your project, if I ever needed to validate `props: Props`, an eventual chained guard would look like this:

```typescript
const isValidProps = both(isWithStyles)
    .and(isWithRouteComponentProps)
    .and(isMyComponentProperties);
```

Not the best of examples, I admit, but in my experience this mix and match of interfaces from different external dependencies is quite common in TypeScript projects -- that is, if you ever want to stay type-safe throughout.

> * When is it an advantage that a chainable guard is itself a type guard?

I just extrapolated `ts-type-guards`'s principle of curried type guards to boolean/logical guard chaining. Everything is a function, and therefore, immediately callable. The advantage is that you don't need an extra build step, e.g. `either(isString).or(isNumber).toGuard()`. Less typing, less error-prone, more stylistic (IMHO, at least).

> * I feel confused by the new implementation of `is`:
>   ```ts
>   const isFive = is(5);
>   ```
>   
>   
>   This wasn't allowed before, because I thought it wasn't reasonable to use the value `5` as a type. Furthermore, `is` is intended to mimic `instanceof`, which throws an error if the right hand side is a primitive value. (Of course we'd rather catch that error at compile time than throw it at runtime.)

I didn't change `is`'s semantics, but I did add some extra typing information, so that `guard` (now `wrap`, or `chainable`, I dunno :p) could inform the correct returned chainable type if needed, e.g. `guard(is(String)): UnionTypeGuard<string>`. But I must confess that I may have not completely understood `is`'s implementation at first.

Recall that the previous signature was `function is<T>(type: Classy<T>): TypeGuard<T>`. This means that it's not sane to pass anything other than a classy-value. This includes classes and functions with prototypes; in other words, the only allowed value type is, effectively, `object`. Yet, a check is made against primitive value (`if (isPrimitive(type)) { ... }`), in which case it returns a contradictory (as in always false) type guard. I didn't understand at the time what was the reasoning behind it, given that e.g. `typeof (new String()) === 'object'` (and therefore, `is(String)(x)` is effectively `x instanceof String`). And that's why I added the `export function is<T extends primitive>(reference: T): PrimitiveTypeGuard<primitive>;` signature: out of confusion. :(

> * I would also like to ask if you could explain the design choices behind all the new types and how they fit together. I'm thinking of `PrimitiveTypeGuard`, `NonPrimitiveTypeGuard`, `UnionTypeGuard`, `ChainableTypeGuard` and `FullyChainableTypeGuard`.

Surely. My bad for not documenting them beforehand.

First, the types that further qualifies a `TypeGuard`:

- `PrimitiveTypeGuard` is the interface describing type guards operating over primitive types. Chains starting with a `PrimitiveTypeGuard` are only allowed to continue with `or`, given that stuff like e.g. `string & Animal` make little to no sense. By pure type conformance, all primitive type guards in the library today are `PrimitiveTypeGuard`s (only because they don't export a `readonly chainable: true` property). Notice, however, that `chainable` is still an optional property that, if given, must always be `false`.
- `NonPrimitiveTypeGuard` is the type of non-primitive type guards. This distinction is made because it may make some sense to work in terms of intersections of interfaces without having to determine a base interface or type alias, e.g. `IAnimal & IMineral`.  `is` should always return a `NonPrimitiveTypeGuard` (but please see my previous comment about its signatures). `isLike` should return a `NonPrimitiveTypeGuard` iff not given a non-primitive value. Conformance is dependent on the existence of a `readonly chainable: true` property that's verifiable at both compile time and runtime.

Note that neither `PrimitiveTypeGuard` nor `NonPrimitiveTypeGuard` are chainable by themselves. Those interfaces were defined in order to allow for type detection when using `guard`.

Now, for the actually chainable interfaces:

- `ChainableTypeGuard` is a basic interface defining the common fields expected for any chainable type guard. Note that it has an optional `and`, exactly because not every chainable type guard is able to be intersectioned with others. This is a catch-all interface; if we cannot know for sure about the presence of `and` (e.g. type erasure, input is just a `TypeGuard`, insufficient type information to determine if the initial guard is a `PrimitiveTypeGuard` or a `NonPrimitiveTypeGuard`), `ChainableTypeGuard` is the type that'll be used as the return of `guard`.  Uses of `and` in this case must pass through a guard first. This interface is provided for type-safety reasons. A `ChainableTypeGuard` is type-compatible with a `TypeGuard`.
- `UnionTypeGuard` is the interface describing all the chainable guard kinds for which it is known that an `and` chain is not possible. This is the type returned by `guard` for any `PrimitiveTypeGuard`. Uses of `and` in this case are prohibited and will generate a compiler error. A `UnionTypeGuard` is type-compatible with a `PrimitiveTypeGuard`.
- `FullyChainableTypeGuard` is the type of a type guard chain that was started with a `NonPrimitiveTypeGuard`, and therefore, can unionize or intersect with any other guard. `and` is guaranteed to be present. A `FullyChainableTypeGuard` is type-compatible with a `NonPrimitiveTypeGuard`.

 Note that `guard` has the following signatures:

```typescript
export function guard<T extends object>(guard: NonPrimitiveTypeGuard<T>): FullyChainableTypeGuard<T>;
export function guard<T extends primitive>(guard: PrimitiveTypeGuard<T>): UnionTypeGuard<T>;
export function guard<T>(guard: TypeGuard<T>): ChainableTypeGuard<T>;
```

> I just pushed some commits which I had forgotten to push. One of them adds `isNothing` and `isSomething`; the rest is just fixing vulnerabilities and upgrading the TypeScript version. I think you should be able to merge those changes fairly easily. Otherwise I'll be happy to help.

Ok!

> By the way, I think your PR should not edit `package-lock.json`. It seems like the changes you made to it were actually made by `npm install`, so I feel obliged to mention `git add -p` – I always use that flag myself! :)

Again, my bad, I'm sorry... :(

> And one last thing: I've been using commit messages starting with a capital letter up until now, so I'd suggest we stick to that convention.

Ok, will follow that convention from now on.

---

Thank you very much for taking your time to revise my PR! Will wait for your comments before proceeding.

SimonAlling commented:
@flisboac I haven't forgotten about this, but I'm very busy these days since I'm doing my master's thesis. I don't know at the moment when I'll have time to discuss these amazing PRs; it could take another 5–6 weeks or so in the worst case. I'd personally rather wait and do it properly than rush through it. Thoughts?

SimonAlling commented:
Hi again, @flisboac! I'm back. I've just read through the entire thread again. Where were we? I'll try to summarize:

* I like this PR very much.
* What should we do with `is`? (You changed its type somewhat, remember?)
* `guard` should be renamed to `wrap`.
* We should export `both` and `either`.
* I made some textual changes above; what should be done for them to be included?

By the way, I have one concern in general that the package might grow "too large". Not from this PR, but down the road. We should keep that in mind; maybe at some point we should split the package up.

flisboac commented:
Hey @SimonAlling ! It has been quite some time, indeed!

Well, this wait was for the better. I've been thinking about my PRs in the meantime, and it occurred to me that, even though the overall idea is good (e.g. guard composition), I'm complicating things too much for little benefit. 

Although I implemented helper functions, the extra step needed to use this fluid API complicates things, both in implementation and use, and could potentially discourage new users. Instead, what would be better is to refrain from customizing guards with "static" functions and instead compose through function calls. It may be too much, but still I worry about the runtime cost of constructing this many functions for each guard when they may in fact never be used in most use cases. You know, "don't pay for what you don't use." :)

In other words, `guard` should be removed; instead, I think I should have introduced only `both` and `either`, but with better (or at the very least, alternative) names: `allOf` and `anyOf` (respectively). Also, I should have dropped the whole distinction between `NonPrimitiveTypeGuard` and `PrimitiveTypeGuard`; the user may be able to make nonsensical unions and intersections in this case, but at least we don't make things more complicated than they should be. Let the language sort the nonsense, either at declaration/definition or use site; it'll break at compile-time in some form anyway.

Also, for the value fallback and type-casting use cases, perhaps a better approach would be to construct a monadic type, and from it you could either get (conditionally or not; effectively, the type-cast), map (to introduce some transformation or cover edge cases), refine (to introduce more rules whilst maintaining the same base type) or check if the value is present. These monadic types can be implemented very efficiently, and would significantly reduce the number of objects created per custom guard. For example, supposing the library provides the following:

```typescript
interface Just<T> {
  get(): T;
  orElse(value: any): T;
}

interface None<T> {
  get(): never;
  orElse<R extends T>(value: R): R;
}

interface MonadFactory<T> {
  (value: T): Just<T>;
  (value: any): None<T>;
}

declare function wrapIf<T>(guard: Guard<T>): MonadFactory<T>;
```

I could use it like the following:

```typescript

const isUserIdentification = anyOf(isString, isNumber);
const maybeUserIdentification= wrapIf(isUserIdentification);

// ...

declare var providedUserId: any;
const name = maybeUserIdentification(providedUserId).orElse("anonymous");
```

Much cleaner and functional.

In any case, I've been writing a toy library to experiment with some new ideas (including the ones I've been discussing with you in my three PRs). Because of that, I don't know what to do with the PRs... Should I close them? (Btw, would you be interested in seeing a (github) gist of it?)

SimonAlling commented:
Hello again! My turn to be sorry for replying slowly.

Yes, I think we should strive not to complicate the library too much. I've even thought about removing some of the code I wrote before you came along.

As for the monad stuff, I do like it a lot, but do you think it should be part of `ts-type-guards`? Maybe that would

A Gist would be interesting to see, I believe. If you think it would be easier to close these PRs and create new ones, feel free to do so!

SimonAlling commented:
Hello again! My turn to be sorry for replying slowly.

Yes, I think we should strive not to complicate the library too much. I've even thought about removing some of the code I wrote before you came along.

Along those lines: Maybe adding stuff that _isn't type guards_ wouldn't align very well with the UNIX principle? What are your thoughts on that? Also, do you think it would be possible and practical to put the very nice monad stuff you suggested in another library that builds upon `ts-type-guards`?

A Gist would be interesting to see, I believe. If you think it would be easier to close these PRs and create new ones, feel free to do so!

flisboac commented:
> Along those lines: Maybe adding stuff that _isn't type guards_ wouldn't align very well with the UNIX principle? What are your thoughts on that? Also, do you think it would be possible and practical to put the very nice monad stuff you suggested in another library that builds upon `ts-type-guards`?

Fair point. For the monad idea, a separate library would be better, and thanks to how TypeScript handles types, there is not even any need for depending on `ts-type-guards` (I mean, given that the only definition needed would be the `Guard` interface, which is just a language concept).

But I still think that there's space for guard composition in `ts-type-guards`, considering that composing guards would yield a new guard covering a subset of the guarded types. I'll redo my PR for that, if you agree with it.

> A Gist would be interesting to see, I believe.

I made a quick gist for the monad idea: https://gist.github.com/flisboac/7d1e94eaaea3629a81f0bd1a063af0b3

(I ran out of energy to document the rest of the file, please bear with me! :D)

Feel free to suggest anything (or edit, if it's even possible).

> If you think it would be easier to close these PRs and create new ones, feel free to do so!

Will do, right now!

--- 

Anyways, thank you for reviewing my (messy) PRs!

SimonAlling commented:
Hi again, @flisboac! I was just about to revisit this PR, because I always liked it; I just found it hard to review so I sort of postponed it over and over. However, the branch seems to have been deleted. I could perhaps reconstruct the content using the info above.
